File: src/supernode/src/main/java/com/alibaba/dragonfly/supernode/repository/PieceState.java
/*
 * Copyright 1999-2017 Alibaba Group.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.alibaba.dragonfly.supernode.repository;

import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.atomic.AtomicInteger;

import com.alibaba.dragonfly.supernode.common.Constants;
import com.alibaba.dragonfly.supernode.common.util.BeanPoolUtil;
import com.alibaba.dragonfly.supernode.common.util.PowerRateLimiter;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PieceState {
    private static final Logger pieceHitLogger = LoggerFactory
        .getLogger("PieceHitLogger");

    private static final Logger logger = LoggerFactory
        .getLogger(PieceState.class);

    private static final ProgressRepository progressRepo = BeanPoolUtil
        .getBean(ProgressRepository.class);
    private static final PowerRateLimiter rateLimiter = BeanPoolUtil
        .getBean(PowerRateLimiter.class);
    private volatile String superCid;
    private Integer pieceSize;
    private ArrayBlockingQueue<String> pieceContainer = new ArrayBlockingQueue<>(50000);

    private volatile int distributedCount = 0;

    public PieceState(Integer pieceSize) {
        this.pieceSize = pieceSize;
    }

    public boolean offerProducer(String cid) {
        if (StringUtils.isBlank(cid)) {
            return false;
        }
        if (superCid == null && cid.startsWith(Constants.SUPER_NODE_CID)) {
            superCid = cid;
        } else {
            synchronized (pieceContainer) {
                if (!pieceContainer.contains(cid)) {
                    if (pieceContainer.offer(cid)) {
                        distributedCount++;
                    }
                } else {
                    logger.warn("cid:{} is already in queue", cid);
                }
            }
        }
        return true;
    }

    public String popProducer(String taskId, String srcCid, int pieceLen) {
        String dstCid = null;

        Set<String> blackSet = progressRepo.getClientBlackInfo(srcCid);

        AtomicInteger clientErrorCount = progressRepo
            .getClientErrorInfo(srcCid);
        if (clientErrorCount == null
            || clientErrorCount.get() <= Constants.FAIL_COUNT_LIMIT) {
            dstCid = tryGetCid(blackSet);
        } else {
            pieceHitLogger.error("srcCid:{} taskId:{} reach error limit",
                srcCid, taskId);
        }

        if (dstCid == null && superCid != null) {
            if (rateLimiter.tryAcquire(pieceLen > 0 ? pieceLen : pieceSize, distributedCount > 1)) {

                dstCid = superCid;
            }
        }

        return dstCid;
    }

    private String tryGetCid(Set<String> blackSet) {
        AtomicInteger load;
        String cid = null;
        String tmpCid;
        int times = pieceContainer.size();
        boolean needOffer;
        while (times-- > 0) {
            needOffer = true;
            tmpCid = pieceContainer.peek();

            if (tmpCid == null) {
                pieceHitLogger.info("peek element from empty queue");
                break;
            }
            if (progressRepo.getServiceDownInfo(tmpCid)) {
                needOffer = false;
            } else {
                AtomicInteger errorCount = progressRepo
                    .getServiceErrorInfo(tmpCid);
                if (errorCount != null
                    && errorCount.get() >= Constants.ELIMINATION_LIMIT) {
                    needOffer = false;
                } else {
                    if (blackSet == null || !blackSet.contains(tmpCid)) {
                        load = progressRepo.getProducerLoad(tmpCid);
                        if (load != null) {
                            if (load.incrementAndGet() <= Constants.PEER_UP_LIMIT) {
                                cid = tmpCid;
                                break;
                            } else {
                                load.decrementAndGet();
                            }
                        } else {
                            needOffer = false;
                        }
                    }
                }
            }
            synchronized (pieceContainer) {
                if (StringUtils.equals(pieceContainer.peek(), tmpCid)) {
                    if (pieceContainer.remove(tmpCid) && needOffer) {
                        pieceContainer.offer(tmpCid);
                    }
                }
            }

        }

        return cid;
    }

    public int getDistributedCount() {
        return distributedCount;
    }

}

File: supernode/daemon/mgr/progress/piece_state.go
/*
 * Copyright The Dragonfly Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package progress

import (
	"github.com/dragonflyoss/Dragonfly/pkg/errortypes"
	"github.com/dragonflyoss/Dragonfly/pkg/stringutils"
	"github.com/dragonflyoss/Dragonfly/pkg/syncmap"

	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"
)

// pieceState maintains the information about
// which peers the piece currently exists on.
type pieceState struct {
	pieceContainer *syncmap.SyncMap
}

// newPieceState returns a new pieceState.
func newPieceState() *pieceState {
	return &pieceState{
		pieceContainer: syncmap.NewSyncMap(),
	}
}

// add a peerID for the corresponding piece which means that
// there is a new peer node that owns this piece.
func (ps *pieceState) add(peerID string) error {
	if stringutils.IsEmptyStr(peerID) {
		return errors.Wrap(errortypes.ErrEmptyValue, "peerID")
	}

	ok, err := ps.pieceContainer.GetAsBool(peerID)
	if err == nil && ok {
		logrus.Warnf("peerID: %s is exist", peerID)
		return nil
	}

	if err != nil && !errortypes.IsDataNotFound(err) {
		return err
	}

	return ps.pieceContainer.Add(peerID, true)
}

func (ps *pieceState) getAvailablePeers() []string {
	return ps.pieceContainer.ListKeyAsStringSlice()
}

func (ps *pieceState) delete(peerID string) error {
	return ps.pieceContainer.Remove(peerID)
}
